spot_driver/spot_driver/calibrated_reregistered_hand_camera_depth_publisher.py

```diff
@@ -0,0 +1,233 @@
+            if param_name in ["camera_matrix_depth", "camera_matrix_rgb", "depth_t_rgb_R"]:
+                self.calibration[param_name] = np.array(self.calibration[param_name]).reshape((3, 3))
+            elif "dim" not in param_name and "name" not in param_name and "undistort" not in param_name:
+                self.calibration[param_name] = np.array(self.calibration[param_name])
+
+        self.cv_bridge = CvBridge()
+
+        raw_depth_topic = f"/{self.calibration['robot_name']}/depth/hand/image"
+        self.node.get_logger().info(f"Creating subscriber to raw depth at {raw_depth_topic}")
+        self.raw_depth_img_sub = self.node.create_subscription(
+            Image, raw_depth_topic, self.republish_registered_depth_callback, 10
+        )
+
+        reregistered_depth_topic = f"/{self.calibration['robot_name']}/{topic_name}"
+        self.node.get_logger().info(f"Creating reregistered depth publisher to {reregistered_depth_topic}")
+        self.reregistered_depth_img_pub = self.node.create_publisher(Image, reregistered_depth_topic, 10)
+
+    def republish_registered_depth_callback(self, msg: Image) -> None:
+        if self.calibration["rgb_image_dim"][0] != 0:
+            raw_depth_image = self.cv_bridge.imgmsg_to_cv2(msg, desired_encoding="passthrough")
+            pointcloud_pcd = self.depth_img_to_pointcloud(raw_depth_image, undistort=self.calibration["undistort"])
+            reregistered_depth_image = self.pointcloud_to_depth_img(pointcloud_pcd)
+            reregistered_depth_msg = msg
+            reregistered_depth_msg.height = self.calibration["rgb_image_dim"][0]
+            reregistered_depth_msg.width = self.calibration["rgb_image_dim"][1]
+            reregistered_depth_msg.header.frame_id = f"{self.calibration['robot_name']}/hand_color_image_sensor"
+
+            reregistered_depth_msg = self.cv_bridge.cv2_to_imgmsg(
+                reregistered_depth_image, encoding="16UC1", header=reregistered_depth_msg.header
+            )
+            self.reregistered_depth_img_pub.publish(reregistered_depth_msg)
+        else:
+            self.node.get_logger().warning(
+                "Cannot republish registered depth image without populating needed parameters."
+            )
+
+    def depth_img_to_pointcloud(
+        self, depth_img: np.ndarray, undistort: bool = False, depth_scale: float = 1000.0, depth_max: float = 10
+    ) -> o3d.t.geometry.PointCloud:
+        depth_img = depth_img.astype(np.float32)
+
+        if undistort:
+            camera_matrix = self.calibration["camera_matrix_depth"]
+            dist_coeffs = np.array(self.calibration["dist_coeffs_depth"])
+            h, w = depth_img.shape[:2]
+            new_camera_matrix, _ = cv2.getOptimalNewCameraMatrix(camera_matrix, dist_coeffs, (w, h), 1, (w, h))
+            map1, map2 = cv2.initUndistortRectifyMap(
+                camera_matrix, dist_coeffs, None, new_camera_matrix, (w, h), cv2.CV_32FC1
+            )
```

Comment:
Does this all need to be recreated each time or can this be done in the constructor?

Reply:
+1

Reply:
```calculate_undistortion_parameters``` now calculates these parameters, called once in the constructor